### PR TITLE
Revert user password requirements

### PR DESCRIPTION
**Pull request by alexhung**
See https://github.com/jfrog/terraform-provider-artifactory/discussions/333

First part of resolving user password issue:
- `password` is optional again
- auto generate password if no value is supplied
- Fix state management so no drift for both cases whether password is supplied or not.

### CODE REVIEW

File: pkg/artifactory/resource_artifactory_user.go [hunk 1]
```diff
@@ -1,19 +1,22 @@
...
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/sethvargo/go-password/password"
 )
 
 type User struct {
 	Name                     string   `json:"name"`
 	Email                    string   `json:"email"`
-	Password                 string   `json:"password"`
+	Password                 string   `json:"password,omitempty"`
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_user.go:19]: `omitempty` is actually useful for this case when `password` is empty and we don't want to send `""` to Artifactory

File: pkg/artifactory/resource_artifactory_user.go [hunk 2]
```diff
@@ -88,10 +91,9 @@ func resourceArtifactoryUser() *schema.Resource {
 			"password": {
 				Type:             schema.TypeString,
 				Sensitive:        true,
-				Required:         true,
-				ValidateDiagFunc: validation.ToDiagFunc(validation.StringIsNotEmpty),
-				Description: "Password for the user. Password validation is not done by the provider and is " +
-					"offloaded onto the Artifactory.",
+				Optional:         true,
+				ValidateDiagFunc: validation.ToDiagFunc(defaultPassValidation),
```
Comment (alexhung) [inline pkg/artifactory/resource_artifactory_user.go:95]: Oh wait, did we decide *not* to validate? I forgot.

Comment (chb0github) [reply pkg/artifactory/resource_artifactory_user.go:95]: Nope. Get out of the password business and fail late

File: pkg/artifactory/resource_artifactory_user.go [hunk 3]
```diff
@@ -144,29 +143,42 @@ func packUser(user User, d *schema.ResourceData) error {
...
-		return fmt.Errorf("no password supplied. Please use any of the terraform random password generators")
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Warning,
+			Summary:  "No password supplied",
+			Detail:   "One will be generated (10 characters with 1 digit, 1 symbol, with upper and lower case letters) and this may fail as your Artifactory password policy can't be determined by the provider.",
+		})
+
+		// Generate a password that is 10 characters long with 1 digit, 1 symbol,
+		// allowing upper and lower case letters, disallowing repeat characters.
+		randomPassword, err := password.Generate(10, 1, 1, false, false)
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_user.go:165]: I guess. It doesn't seem very functional to me. I would have done something like this:

```go
password.Generate(
     min(10),
     digit(1),
     symbol(1),
     upper(),
     lower()
)
```
And then the generator literally just executes each function

But, whatever

File: pkg/artifactory/resource_artifactory_user_test.go [hunk 4]
```diff
@@ -73,9 +73,9 @@ func TestAccUserPasswordNotChangeWhenOtherAttributesChangeGH340(t *testing.T) {
 func TestAccUser_basic(t *testing.T) {
 	const userBasic = `
 		resource "artifactory_user" "%s" {
-			name  	= "the.dude%d"
+			name  	= "dummy_user%d"
```
Comment (chb0github) [inline pkg/artifactory/resource_artifactory_user_test.go:76]: You don't like?!

File: pkg/artifactory/resource_artifactory_user.go (hunk 3 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_user.go:165]: Take that up with https://github.com/sethvargo/go-password

File: pkg/artifactory/resource_artifactory_user_test.go (hunk 4 above)
Comment (alexhung) [reply pkg/artifactory/resource_artifactory_user_test.go:76]: Just trying to be consistent with the other tests 😄

**Review by chb0github — DISMISSED** [review]

**Review by chb0github — DISMISSED** [review]

**Review by chb0github — APPROVED** [review]

**Pull request merged**